src/test/java/net/snowflake/client/jdbc/FileUploaderExpandFileNamesTest.java

```diff
@@ -126,4 +138,74 @@ public int read() throws IOException {
     assertEquals("dummy_dest_file_name", config.getDestFileName());
     assertEquals(expectedThrowCount, throwCount);
   }
+
+  /**
+   * We have N jobs expanding files with exclusive pattern, processing them and deleting. Expanding
+   * the list should not cause the error when file of another pattern is deleted which may happen
+   * when FileUtils.listFiles is used.
+   *
+   * <p>Fix available after version 3.16.1.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testFileListingDoesNotFailOnMissingFilesOfAnotherPattern() throws Exception {
+    folder.newFolder("TestFiles");
+    String folderName = folder.getRoot().getCanonicalPath();
+
+    int filePatterns = 10;
+    int filesPerPattern = 100;
+    IntStream.range(0, filesPerPattern * filePatterns)
+        .forEach(
+            id -> {
+              try {
+                File file =
+                    new File(
+                        folderName
+                            + localFSFileSep
+                            + "foo"
+                            + id % filePatterns
+                            + "-"
+                            + UUID.randomUUID());
+                assertTrue(file.createNewFile());
+              } catch (IOException e) {
+                throw new RuntimeException(e);
+              }
+            });
+
+    ExecutorService executorService = Executors.newFixedThreadPool(filePatterns / 3);
+    List<Future<Set<String>>> futures = new ArrayList<>();
+    for (int i = 0; i < filePatterns; ++i) {
+      String[] locations = {
+        folderName + localFSFileSep + "foo" + i + "*",
+      };
+      Future<Set<String>> future =
+          executorService.submit(
+              () -> {
+                try {
+                  Set<String> strings = SnowflakeFileTransferAgent.expandFileNames(locations, null);
+                  strings.forEach(
+                      fileName -> {
+                        try {
+                          File file = new File(fileName);
+                          Files.delete(file.toPath());
+                        } catch (IOException e) {
+                          throw new RuntimeException(e);
+                        }
+                      });
+                  return strings;
+                } catch (SnowflakeSQLException e) {
+                  throw new RuntimeException(e);
+                }
+              });
+      futures.add(future);
+    }
+    executorService.shutdown();
+    assertTrue(executorService.awaitTermination(60, TimeUnit.SECONDS));
```

Comment:
Do we need 60 seconds? Maybe it can be shorter?

Reply:
It does not matter - it's much faster but I prefer to leave it like that